flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SortMergeResultPartition.java

```diff
@@ -0,0 +1,360 @@
+		this.writeBuffer = MemorySegmentFactory.allocateUnpooledOffHeapMemory(networkBufferSize);
+	}
+
+	@Override
+	protected void releaseInternal() {
+		synchronized (lock) {
+			isFinished = true; // to fail writing faster
+
+			// delete the produced file only when no reader is reading now
+			if (readers.isEmpty()) {
+				if (resultFile != null) {
+					resultFile.deleteQuietly();
+					resultFile = null;
+				}
+			}
+		}
+	}
+
+	@Override
+	public void emitRecord(ByteBuffer record, int targetSubpartition) throws IOException {
+		emit(record, targetSubpartition, DataType.DATA_BUFFER);
+	}
+
+	@Override
+	public void broadcastRecord(ByteBuffer record) throws IOException {
+		broadcast(record, DataType.DATA_BUFFER);
+	}
+
+	@Override
+	public void broadcastEvent(AbstractEvent event, boolean isPriorityEvent) throws IOException {
+		Buffer buffer = EventSerializer.toBuffer(event, isPriorityEvent);
+		try {
+			ByteBuffer serializedEvent = buffer.getNioBufferReadable();
+			broadcast(serializedEvent, buffer.getDataType());
+		} finally {
+			buffer.recycleBuffer();
+		}
+	}
+
+	private void broadcast(ByteBuffer record, DataType dataType) throws IOException {
+		for (int channelIndex = 0; channelIndex < numSubpartitions; ++channelIndex) {
+			record.rewind();
+			emit(record, channelIndex, dataType);
+		}
+	}
+
+	private void emit(ByteBuffer record, int targetSubpartition, DataType dataType) throws IOException {
+		checkInProduceState();
+
+		SortBuffer sortBuffer = getSortBuffer();
+		if (sortBuffer.append(record, targetSubpartition, dataType)) {
+			return;
+		}
+
+		if (!sortBuffer.hasRemaining()) {
+			// the record can not be appended to the free sort buffer because it is too large
+			releaseCurrentSortBuffer();
+			writeLargeRecord(record, targetSubpartition, dataType);
+			return;
+		}
+
+		flushCurrentSortBuffer();
+		emit(record, targetSubpartition, dataType);
+	}
+
+	private void releaseCurrentSortBuffer() {
+		if (currentSortBuffer != null) {
+			currentSortBuffer.release();
+			currentSortBuffer = null;
+		}
+	}
+
+	private SortBuffer getSortBuffer() {
+		if (currentSortBuffer != null) {
+			return currentSortBuffer;
+		}
+
+		currentSortBuffer = new PartitionSortedBuffer(bufferPool, numSubpartitions, networkBufferSize);
+		return currentSortBuffer;
+	}
+
+	private void flushCurrentSortBuffer() throws IOException {
+		if (currentSortBuffer == null || !currentSortBuffer.hasRemaining()) {
+			releaseCurrentSortBuffer();
+			return;
+		}
+
+		currentSortBuffer.finish();
+		PartitionedFileWriter fileWriter = getPartitionedFileWriter();
+
+		while (currentSortBuffer.hasRemaining()) {
+			BufferWithChannel bufferWithChannel = currentSortBuffer.copyData(writeBuffer);
+			Buffer buffer = bufferWithChannel.getBuffer();
+			int subpartitionIndex = bufferWithChannel.getChannelIndex();
+
+			writeCompressedBufferIfPossible(buffer, fileWriter, subpartitionIndex);
+			updateStatistics(buffer, subpartitionIndex);
```

Comment:
nit: `updateStatistics` could be covered inside `writeCompressedBufferIfPossible` since both `#writeLargeRecord` and `#flushCurrentSortBuffer`  have the same usages, then it can deduplicate the code a bit.